### ISSUE TITLE
Arguments to Main

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#977]
Add support for reading arguments passed to main.

**Comment by mlutze** [comment]
Should arguments be taken as an array or as a list?

**Comment by magnus-madsen** [comment]
Good question. Given that Flix has both, what do you think?

**Comment by mlutze** [comment]
From a look at other functional languages, it seems that array is the most common choice. OCaml, F#, and Scala all take arguments as an array. Haskell takes arguments as a list, however.

I personally lean toward list. There's no reason we should be mutating the arguments, and a list seems to lend itself more naturally to parsing. In the (I would guess rare) case where there are many command line arguments and random access to them is needed, the user could convert to a list themself.

**Comment by mlutze** [comment]
While we're adjusting the signature, should we also declare the function `Impure`?

**Comment by magnus-madsen** [comment]
I think there are several interesting questions here:

- What arguments should main take?
- What should main return?
- What should the effect of main be?
- "Where" should main reside?

At the moment all of these choices where made some what haphazardly. 

One thing I have found very useful is that if main returns a value then that value is printed to the console. I have also found it useful that you dont have to write a long signature for a main that takes no arguments. But I am not sure if such features can still continue to be supported if we come up with a more strict and correct definition of main.

**Comment by magnus-madsen** [comment]
If we are in the process of re-designing main, we may wonder if it should be given *extra* arguments, e.g.:

- References to standard in, standard out, and standard err?
- A map of the local environmental variables?
- The "name" of the OS?
- The line separator of the OS?
- The number of processors?
- The user name?
- The current working directory?
- ...

Obviously they would not have to be individual arguments-- they could be placed in some suitable record.

Also, I am not saying this is a good idea. It is just food for thought.

**Comment by magnus-madsen** [comment]
Another question:

What should main return? Unit? An int indicating exit status? A result (so that it can fail)? An arbitrary value that we turn to a string and print?

**Comment by magnus-madsen** [comment]
I have had this (crazy?) idea that main should take an extensible record:

For example:

```
def main(w: World): Int = 0
```

where:

```
type alias World = {
  args: Array[String],
  currentWorkingDirectory: String,
  console: { stdIn: ..., stdOut: ..., stdErr: ...},
  platform: ...,
  hardware: { processors: Int  }
}
```

and so forth.

We can then use this like:

```
def main(w: World): Int = foo(w)

def foo(r: {hardware: {processors: Int | _} | _): Int =r.hardware.processors
```

(If we allow rows to be implicitly open then we could omit the `| _`)

Thoughts?

**Comment by mlutze** [comment]
I like the "World" idea. I wonder if it should be an extensible record so that in case we add fields in the future, it will be backward-compatible.

I think it is important to be able to control the exit code without needing to make an effectful call like `exit(1)`.

I don't like arbitrary printing since a `--quiet` program for example would be more complex to implement. I think console output should be explicit.

So I suggest main return either an Int or something like `ExitCode(Int)` for a bit more safety.

**Comment by magnus-madsen** [comment]
> I like the "World" idea. I wonder if it should be an extensible record so that in case we add fields in the future, it will be backward-compatible.

I agree. It does require some work though. The only concern is that I wonder whether we would have to put too much stuff in there.


> I think it is important to be able to control the exit code without needing to make an effectful call like `exit(1)`.

Yes, definitely.

> I don't like arbitrary printing since a `--quiet` program for example would be more complex to implement. I think console output should be explicit.

Yeah, that is mostly a convenience we have been using. I don't think main should print anything by default.

> So I suggest main return either an Int or something like `ExitCode(Int)` for a bit more safety.

I think main should just return an Int.

(I guess it could also return a string and then we wrap that in a printline).

**Comment by magnus-madsen** [comment]
You know, if we are not bound by tradition, there is no reason why we could not "overload" main. For example:

```
def main(): Unit // OK
def mainWithReturn(): Int // OK
def mainWithWorld(w: World): Int // OK
```

And you would have at most one of these. 

I don't know, maybe this is too far of the beaten path.
